borgmatic/commands/completion.py

```diff
@@ -55,3 +65,153 @@ def bash_completion():
             '\ncomplete -o bashdefault -o default -F complete_borgmatic borgmatic',
         )
     )
+
+
+# fish section
+
+
+def has_file_options(action: Action):
+    return action.metavar in (
+        'FILENAME',
+        'PATH',
+    ) or action.dest in ('config_paths',)
+
+
+def has_choice_options(action: Action):
+    return action.choices is not None
+
+
+def has_required_param_options(action: Action):
+    return (
+        action.required is True
+        or action.nargs
+        in (
+            '+',
+            '*',
+        )
+        or '--archive' in action.option_strings
+        or action.metavar in ('PATTERN', 'KEYS', 'N')
+        or (action.type is not None and action.default is None)
+    )
+
+
+def has_exact_options(action: Action):
+    return (
+        has_file_options(action) or has_choice_options(action) or has_required_param_options(action)
+    )
+
+
+def exact_options_completion(action: Action):
+    '''
+    Given an argparse.Action instance, return a completion invocation
+    that forces file completion or options completion, if the action
+    takes such an argument and was the last action on the command line.
+
+    Otherwise, return an empty string.
+    '''
+
+    if not has_exact_options(action):
+        return ''
+
+    args = ' '.join(action.option_strings)
+
+    if has_file_options(action):
+        return f'''\ncomplete -c borgmatic -Fr -n "__borgmatic_last_arg {args}"'''
+
+    if has_choice_options(action):
+        return f'''\ncomplete -c borgmatic -f -a '{' '.join(map(str, action.choices))}' -n "__borgmatic_last_arg {args}"'''
+
+    if has_required_param_options(action):
+        return f'''\ncomplete -c borgmatic -x -n "__borgmatic_last_arg {args}"'''
+
+    raise RuntimeError(
+        f'Unexpected action: {action} passes has_exact_options but has no choices produced'
+    )
+
+
+def dedent_strip_as_tuple(string: str):
+    return (dedent(string).strip('\n'),)
```

Comment:
So here's another thing that might seem like overkill: Even though all of these nicely separated utility functions are covered by the big ol' integration / end-to-end tests you added, IMO they should also have unit test coverage. Part of that is because they _are_ so nicely separated and atomic—you can pretty easily pass in a single value (or a mock) and test that each function returns what it's supposed to.

In cases where a function really has no logic to speak of, a simple "does not raise" unit test would be sufficient IMO.

And I still believe though that you don't need any unit testing on `fish_completion()` itself. The cost/benefit there just wouldn't be worth it.

Reply:
I think I did this? Not sure if the tests I wrote were quite what you were looking for.